Results_Engine/Query/Group.cs

```diff
@@ -67,6 +70,108 @@ public static partial class Query
+
+        [Description("Group results by all Scenario type properties of the object.")]
+        [Input("results", "Results to group.")]
+        [Output("groupedResults", "Results grouped as List of Lists.")]
+        public static List<List<T>> GroupByScenario<T>(this IEnumerable<T> results) where T : IResult
+        {
+            if (results == null || !results.Any())
+                return new List<List<T>>();
+
+            return GroupByMultipleSelectors(results, IdentifierFunctions(results.First().GetType(), typeof(ScenarioIdentifierAttribute)));
+        }
+
+        /***************************************************/
+
+        [Description("Group results by Object identifiers of the object.")]
+        [Input("results", "Results to group.")]
+        [Output("groupedResults", "Results grouped as List of Lists.")]
+        public static List<List<T>> GroupByObjectIdentifiers<T>(this IEnumerable<T> results) where T : IResult
+        {
+            if (results == null || !results.Any())
+                return new List<List<T>>();
+
+            return GroupByMultipleSelectors(results, IdentifierFunctions(results.First().GetType(), typeof(ObjectIdentifierAttribute)));
+        }
+
+        /***************************************************/
+
+        [Description("Group results by all identifiers of the object.")]
+        [Input("results", "Results to group.")]
+        [Output("groupedResults", "Results grouped as List of Lists.")]
+        public static List<List<T>> GroupByAllIdentifiers<T>(this IEnumerable<T> results) where T : IResult
+        {
+            if (results == null || !results.Any())
+                return new List<List<T>>();
+
+            return GroupByMultipleSelectors(results, IdentifierFunctions(results.First().GetType(), typeof(IdentifierAttribute)));
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        [Description("Groups by multiple selectors by contiously grouping by each provided selector.")]
+        private static List<List<T>> GroupByMultipleSelectors<T>(this IEnumerable<T> results, IEnumerable<Func<object, object>> funcs)
+        {
+            IEnumerable<IEnumerable<T>> listGroup = new List<IEnumerable<T>>() { results };
+            foreach (var func in funcs)
+                listGroup = listGroup.Select(x => x.GroupBy(y => func(y))).SelectMany(x => x);
+
+            return listGroup.Select(x => x.ToList()).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Gets a list of compiled selectors corresponding to all properties with the attribute type on the type. Allows the attribute type to be defined on an interface of the object type.")]
+        private static List<Func<object, object>> IdentifierFunctions(Type type, Type attributeType)
```

Comment:
```suggestion
        private static List<Func<object, object>> GetIdentifierGetters(Type objectType, Type identifierAttributeType)
```

Reply:
Can change to IdentifierGetters, but will skipe the initial `Get` as we have tended to do in the query methods of the Engine :)